routes/resources.js

```diff
@@ -10,27 +10,44 @@ var Schemas = require('../generators/schemas');
 
 module.exports = function (app, model, Implementation, integrator, opts) {
   var modelName = Implementation.getModelName(model);
-  var schema = Schemas.schemas[Implementation.getModelName(model)];
+  var schema = Schemas.schemas[modelName];
+
+  function tryToAddSmartFieldToRecord(record, fieldObject, modelName) {
+    var smartFieldValue;
+    if (fieldObject.get || fieldObject.value) {
+      if (fieldObject.value) {
+        logger.warn('DEPRECATION WARNING: In your ' + modelName +
+          ' Model, Smart Field value method is deprecated. ' +
+          'Please use get method instead. ');
+        smartFieldValue = fieldObject.value(record);
+      } else {
+        smartFieldValue = fieldObject.get(record);
+      }
+      if (smartFieldValue && _.isFunction(smartFieldValue.then)) {
+        return smartFieldValue.then(function (value) {
+          record[fieldObject.field] = value;
+        });
+      } else {
+        record[fieldObject.field] = smartFieldValue;
+      }
+    } else if (_.isArray(fieldObject.type) && !fieldObject.isVirtual) {
+      record[fieldObject.field] = [];
+    }
+  }
 
   function injectSmartFields(record) {
     return P.each(schema.fields, function (field) {
       if (!record[field.field]) {
-        if (field.get || field.value) {
-          if (field.value) {
-            logger.warn('DEPRECATION WARNING: Smart Field value method is ' +
-              'deprecated. Please use get method instead.');
-          }
-
-          var value = field.get ? field.get(record) : field.value(record);
-          if (value && _.isFunction(value.then)) {
-            return value.then(function (value) {
-              record[field.field] = value;
-            });
-          } else {
-            record[field.field] = value;
-          }
-        } else if (_.isArray(field.type)) {
-          record[field.field] = [];
+        return tryToAddSmartFieldToRecord(record, field, modelName);
+      } else if (field.reference) {
+        // Inject Smart Field in BelongsTo Fields
+        var belongsToModelName = field.reference.split('.')[0];
+        var referenceFieldSchema = Schemas.schemas[belongsToModelName];
+        if (referenceFieldSchema) {
+          _.each(referenceFieldSchema.fields, function (fieldObject) {
```

Comment:
If the Smart Fields return Promises, the current implementation does not work. I think you should wait for Promises to be resolved before sending the result to the front.

Here is an example of "asynchronous" Smart Fields to debug the current implementation:

```
'use strict';
const _ = require('lodash');
const models = require('../models');
const Liana = require('forest-express-sequelize');

Liana.collection('user', {
  fields: [{
    field: 'fullName',
    type: 'String',
    value: function (user) {
      return models.user.findById(2).then(() => {
        if (user.firstName && user.lastName) {
          return `${user.firstName} ${user.lastName}`;
        } else {
          return null;
        }
      });
    },
    search: function (query, search) {
      var s = models.sequelize;
      var split = search.split(' ');

      var searchCondition = s.and(
        { firstName: { $ilike: split[0] }},
        { lastName: { $ilike: split[1] }}
      );

      let searchConditions = _.find(query.where.$and, '$or');
      searchConditions.$or.push(searchCondition);
    }
  }, {
    field: 'fullName2',
    type: 'String',
    value: function (user) {
      return models.user.findById(2).then(() => {
        if (user.firstName && user.lastName) {
          return `${user.firstName} ${user.lastName}`;
        } else {
          return null;
        }
      });
    },
    search: function (query, search) {
      var s = models.sequelize;
      var split = search.split(' ');

      var searchCondition = s.and(
        { firstName: { $ilike: split[0] }},
        { lastName: { $ilike: split[1] }}
      );

      let searchConditions = _.find(query.where.$and, '$or');
      searchConditions.$or.push(searchCondition);
    }
  }]
});
```